### PR TITLE
fix(security): Rate-limit and size-limit peer transaction ID messages

### PR DESCRIPTION
**Pull request by teor2345**
## Motivation

Zebra can send a transaction gossip to half its peers for every transaction it verifies. This could put a lot of load on the network if transaction rates increase.

This helps with the following tickets, but doesn't completely resolve any of them:
- peer set load: #6506
- rate-limit some messages: #2153
- `inv` denial of service: #3217
- more transactions on the network after the deployment of #6556 

### Specifications

Zcash network messages are limited to 2MB.

Gossip behaviour is implementation-specific.

The larger wide transaction ID inventory gossip message size of 68 bytes is documented in:
https://zips.z.cash/zip-0239#deployment

### Complex Code or Requirements

This code modifies a channel shared between concurrent async tasks.

## Solution

Fixes - network message size:
- [Split MAX_INV_IN_RECEIVED_MESSAGE and MAX_TX_INV_IN_SENT_MESSAGE](https://github.com/ZcashFoundation/zebra/pull/6625/commits/456a68518b853d2352364e6666e12a7d653bdc4d)
    - After ZIP-239 in NU5, inventory IDs can now be around twice the size, so I halved this limit for sent messages
    - Received messages are still limited by the protocol message limit
    - I also submitted the same change to `zcashd`: https://github.com/zcash/zcash/pull/6643
    - This limit shouldn't be reached with the default config, because the mempool is limited to 8,000 transactions
- [Enforce `MAX_TX_INV_IN_SENT_MESSAGE` in the network layer for gossips and responses](https://github.com/ZcashFoundation/zebra/pull/6625/commits/0d86213ff57d116bffd5bbb6da5a0056c5a7b2ab)
    - The parser already enforces it for requests from peers)

Fixes - network message rate-limits:
- [Combine multiple transaction updates into a single gossip & rate-limit gossips](https://github.com/ZcashFoundation/zebra/commit/5a3681e5cfe58f87ff928dd0b5c4dbe35e2542a3)
    - At the moment, peers can trigger lots of broadcast gossips, by sending transactions that are quick to verify
    - This also fixes some bugs:
        - Some transaction gossips could be dropped if they verify close together
        - Zebra logs each transaction gossip, which is a usability problem
        - In rare cases, Zebra could produce invalid gossips with too many transactions, and generate a network error (which was ignored)
- [Rate-limit block gossips](https://github.com/ZcashFoundation/zebra/commit/c18e993710fce7cc92b808ce31b3e09e0dcaa3c1)
    - This rate limit is unlikely to be reached, because blocks are usually 75 seconds apart
    - Zebra logs each block gossip, which could be a usability problem if blocks arrive close together

Docs:
- [Fix usage documentation for Message::Tx](https://github.com/ZcashFoundation/zebra/pull/6625/commits/e8e0df239a56a90eeda44af5bc8ebf9e27e21144)

Testing:
- [Fix mempool_transaction_expiration gossip test timings](https://github.com/ZcashFoundation/zebra/commit/e426f63a115543c15e749db61619b2b9b7909a2f)

## Review

This is a routine security fix. It doesn't have to go into the next release, but if we have time to review it, it would be nice to get it in.

### Reviewer Checklist

  - [ ] Will the PR name make sense to users?
    - [ ] Does it need extra CHANGELOG info? (new features, breaking changes, large changes)
  - [ ] Are the PR labels correct?
  - [ ] Does the code do what the ticket and PR says?
    - [ ] Does it change concurrent code, unsafe code, or consensus rules?
  - [ ] How do you know it works? Does it have tests?

## Follow Up Work

Rate-limit other kinds of network messages if they become an issue.

### CODE REVIEW

**Comment by teor2345** [comment]
After this change, new blocks and transactions are gossiped immediately, but then the *next* gossip is delayed for 6 seconds, and any pending changes are combined:
```
2023-05-07T23:44:14.798961Z  INFO {zebrad="e426f63" net="Main"}: zebrad::components::mempool::gossip: sending mempool transaction broadcast request=AdvertiseTransactionIds(4) changes=1
2023-05-07T23:44:15.038166Z  INFO {zebrad="e426f63" net="Main"}:{peer=Out("162.19.139.183:8233")}:msg_as_req{msg="inv"}:inbound:download_and_verify{hash=000000000067b02b015bccab67b49a0979145b9b685076c2897343706cd8e9e1}: zebrad::components::inbound::downloads: downloaded and verified gossiped block height=Height(2078266)
2023-05-07T23:44:25.046703Z  INFO {zebrad="e426f63" net="Main"}: zebrad::components::mempool::gossip: sending mempool transaction broadcast request=AdvertiseTransactionIds(7) changes=4
2023-05-07T23:44:31.048226Z  INFO {zebrad="e426f63" net="Main"}: zebrad::components::mempool::gossip: sending mempool transaction broadcast request=AdvertiseTransactionIds(3) changes=3
2023-05-07T23:44:37.049697Z  INFO {zebrad="e426f63" net="Main"}: zebrad::components::mempool::gossip: sending mempool transaction broadcast request=AdvertiseTransactionIds(3) changes=2
2023-05-07T23:44:43.050632Z  INFO {zebrad="e426f63" net="Main"}: zebrad::components::mempool::gossip: sending mempool transaction broadcast request=AdvertiseTransactionIds(1) changes=1
2023-05-07T23:44:49.052259Z  INFO {zebrad="e426f63" net="Main"}: zebrad::components::mempool::gossip: sending mempool transaction broadcast request=AdvertiseTransactionIds(1) changes=1
2023-05-07T23:45:09.622233Z  INFO {zebrad="e426f63" net="Main"}: zebrad::components::mempool::gossip: sending mempool transaction broadcast request=AdvertiseTransactionIds(1) changes=1
```

**Comment by teor2345** [comment]
@mergifyio update

**Comment by teor2345** [comment]
@mergifyio update

File: zebra-network/src/peer/connection.rs [hunk 1]
```diff
@@ -1351,11 +1373,30 @@ where
...
+                //
+                // In most cases, we try to split over-sized responses into multiple network-layer
+                // messages. But we are unlikely to reach this limit with the default mempool
+                // config, so a response like this could indicate a network amplification attack.
+                //
+                // If there are thousands of transactions in the mempool, letting peers know the
+                // exact transactions we have isn't that important, so it's ok to drop arbitrary
+                // transaction hashes from our response.
+                if hashes.len() > max_tx_inv_in_message {
+                    debug!(inv_count = ?hashes.len(), ?MAX_TX_INV_IN_SENT_MESSAGE, "unusually large transaction ID reponse");
```
Comment (teor2345) [inline zebra-network/src/peer/connection.rs]:
```suggestion
                    debug!(inv_count = ?hashes.len(), ?MAX_TX_INV_IN_SENT_MESSAGE, "unusually large transaction ID response");
```

**Comment by teor2345** [comment]
We're seeing an increased volume of transactions on the network after the ZIP-317 deployment, so this seems like a high priority to get fixed.

**Comment by teor2345** [comment]
@mergifyio update

**Review by oxarbitrage — APPROVED** [review]

**Pull request merged**